### PR TITLE
[SPARK-20044][UI] Support Spark UI behind front-end reverse proxy using a path prefix Revert proxy url

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Allow to run the Spark web UI behind a reverse proxy with URLs prefixed by a context root, like www.mydomain.com/spark. In particular, this allows to access multiple Spark clusters through the same virtual host, only distinguishing them by context root, like www.mydomain.com/cluster1, www.mydomain.com/cluster2, and it allows to run the Spark UI in a common cookie domain (for SSO) with other services.

### Why are the changes needed?

This PR is to take over https://github.com/apache/spark/pull/17455.
After changes, Spark allows showing customized prefix URL in all the `href` links of the HTML pages.

### Does this PR introduce _any_ user-facing change?

Yes, all the links of UI pages will be contains the value of `spark.ui.reverseProxyUrl` if it is configurated.
### How was this patch tested?

New HTML Unit tests in MasterSuite
Manual UI testing for master, worker and app UI with an nginx proxy
Spark config:
```
spark.ui.port 8080
spark.ui.reverseProxy=true
spark.ui.reverseProxyUrl=/path/to/spark/
```
nginx config:
```
server {
    listen 9000;
    set $SPARK_MASTER http://127.0.0.1:8080;
    # split spark UI path into prefix and local path within master UI
    location ~ ^(/path/to/spark/) {
        # strip prefix when forwarding request
        rewrite /path/to/spark(/.*) $1  break;
        #rewrite /path/to/spark/ "/" ;
        # forward to spark master UI
        proxy_pass $SPARK_MASTER;
        proxy_intercept_errors on;
        error_page 301 302 307 = @handle_redirects;
    }
    location @handle_redirects {
        set $saved_redirect_location '$upstream_http_location';
        proxy_pass $saved_redirect_location;
    }
}
```

### CODE REVIEW

**Comment by gengliangwang** [comment]
This PR is to take over #17455.

**Review by IngoSchuster — APPROVED** [review]

**Comment by tgravescs** [comment]
the jira for this is resolved as won't fix, if we are working on it can you reopen the jira.

**Comment by gengliangwang** [comment]
@tgravescs Thanks for the reminder.
This is still in progress and I don't think the current changes will work.
I found that the link for "/stages" will become "$prefix_url/proxy/$app_id/stages" after the changes. The reverse proxy should strip the "$prefxi_url" part and send the rest of the URL "proxy/$app_id/stages" to the driver, which won't be handled. The current changes only works for the workers/application links.

I will update the JIRA after investigation.

**Comment by gatorsmile** [comment]
What is the status of this PR?

**Comment by gengliangwang** [comment]
This is now ready for review.
cc @tgravescs @vanzin @ajbozarth @okoethibm

**Comment by gengliangwang** [comment]
cc @srowen  @HeartSaVioR  @cloud-fan as well

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala [hunk 1]
```diff
@@ -276,7 +276,12 @@ private[deploy] class Worker(
     master = Some(masterRef)
     connected = true
     if (reverseProxy) {
-      logInfo(s"WorkerWebUI is available at $activeMasterWebUiUrl/proxy/$workerId")
+      logInfo("WorkerWebUI is available at %s/proxy/%s".format(
+        activeMasterWebUiUrl.stripSuffix("/"), workerId))
+       // if reverseProxyUrl is not set, then we continue to generate relative URLs
+       // starting with "/" throughout the UI and do not use activeMasterWebUiUrl
+       val proxyUrl = conf.get(UI_REVERSE_PROXY_URL.key, "").stripSuffix("/")
+       System.setProperty("spark.ui.proxyBase", proxyUrl)
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala:284]: shall we mention that the `/proxy/...` will be padded in `UIUtils.makeHref`?

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Comment by drauschenbach** [comment]
I don't understand this statement:

>  This setting affects all the workers and application UIs running in the cluster and must be set identically on all the workers, drivers and masters.

In an HA Spark Standalone setup, there are always two Spark Masters, and config files are all static, because that is the model. Workers switch from one master to the other during a failover. So how can they be expected to contain a static reference to one or the other master in their static config?

I expose both Spark Masters in my Nginx config (/spark-master.0/, /spark-master.1/), and I'm struggling to figure out how to roll out this change.

**Comment by gengliangwang** [comment]
@drauschenbach We can treat the prefix as the ID of one cluster. For example, let's say the prefix is "cluster1" and the reverse proxy is nginx, and both master 0 and master 1 are configurated with the prefix "/cluster1".
Before failover, "/cluster1" maps to IP of master0. Accessing master URL starting with "/cluster1" will redirect to master 0. All the shown worker UI are with prefix as well, so that accessing worker URL will go to nginx first, and then master 0, and then the exact worker.
After failover, I think you can try just map "/cluster1" to the IP of master1 in nginx.

**Pull request closed without merging**